### PR TITLE
[FLINK-20613][table] Update TableResult.collect() to the new type system

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This updates `TableResult.collect()` to the new type system and thus reduces type incompatibilities. It is now possible to fully stream all records either `fromDataStream` or `Table` to `collect()`.

## Brief change log

See commit messages.

- During translation a `collect()` behaves like `toDataStream` thus like a regular sink.
- The conversion is done by the `ExternalSerializer`.

## Verifying this change

This change added tests and can be verified as follows: `DataStreamJavaITCase` and all other tests that use `collect()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/connectors/DynamicSinkUtils.java [hunk 1]
```diff
@@ -75,6 +80,49 @@
...
+    /** Temporary solution until we drop legacy types. */
+    private static DataType fixCollectDataType(
+            DataTypeFactory dataTypeFactory, ResolvedSchema schema) {
+        final DataType fixedDataType =
+                DataTypeUtils.transform(
+                        dataTypeFactory,
+                        schema.toSourceRowDataType(),
+                        TypeTransformations.legacyRawToTypeInfoRaw(),
+                        TypeTransformations.legacyToNonLegacy());
+        // TODO erase the conversion class earlier when dropping legacy code
```
Comment (Airblader) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/connectors/DynamicSinkUtils.java:122]: Is this a TODO for a specific issue?

**Review by Airblader — APPROVED** [review]
Approving already since the comments are minor.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/connectors/CollectDynamicSink.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                    final CollectStreamSink<RowData> sink =
+                            new CollectStreamSink<>(inputStream, factory);
+                    return sink.name("Collect table sink");
+                };
+    }
+
+    @Override
+    public DynamicTableSink copy() {
+        final CollectDynamicSink copy = new CollectDynamicSink(tableIdentifier, consumedDataType);
+        copy.iterator = iterator;
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/connectors/CollectDynamicSink.java:114]: True, but the copy method is only called during optimization phase (not before runtime provider getter). So it should not be a problem for this special sink. I will add a comment.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/connectors/DynamicSinkUtils.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/connectors/DynamicSinkUtils.java:122]: I created FLINK-22321

**Pull request closed without merging**